fix(android): Fixes web-view handling url (#10147)

prabudevarrajan opened the pull request:
<!--
We, the rest of the NativeScript community, thank you for your
contribution! 
To help the rest of the community review your change, please follow the instructions in the template.
-->

<!-- PULL REQUEST TEMPLATE -->
<!-- (Update "[ ]" to "[x]" to check a box) -->

## PR Checklist

- [ ] The PR title follows our guidelines: https://github.com/NativeScript/NativeScript/blob/master/tools/notes/CONTRIBUTING.md#commit-messages.
- [ ] There is an issue for the bug/feature this PR is for. To avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it.
- [ ] You have signed the [CLA](http://www.nativescript.org/cla).
- [ ] All existing tests are passing: https://github.com/NativeScript/NativeScript/blob/master/tools/notes/DevelopmentWorkflow.md#running-unit-tests-application.
- [ ] Tests for the changes are included - https://github.com/NativeScript/NativeScript/blob/master/tools/notes/WritingUnitTests.md.

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

## What is the new behavior?
<!-- Describe the changes. -->

Fixes/Implements/Closes #[Issue Number].

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

<!-- 
BREAKING CHANGES:


[Describe the impact of the changes here.]

Migration steps:
[Provide a migration path for existing applications.]
-->

NathanWalker commented on packages/core/ui/web-view/index.android.ts:
```diff
@@ -2,6 +2,7 @@ import { disableZoomProperty, WebViewBase, WebViewClient } from './web-view-comm
 import { Trace } from '../../trace';
 import { knownFolders } from '../../file-system';
 import { openUrl } from '../../utils';
+import * as application from '../../application';
```
Instead of importing `application` here, you can import from utils. We've been trying to pair down some circular imports throughout core over time and `application` contains some which would be better to avoid here for bundling.

Instead, do this:
```
import { openUrl, ad } from '../../utils';
```

## For Clarity

`ad` when importing within core relatively is short for `android` to avoid bumping up against the `android` global. It is exported from the main barrel for consumption as:

```
import { Utils } from '@nativescript/core'

// for example:
Utils.android.getApplicationContext()
```

NathanWalker commented on packages/core/ui/web-view/index.android.ts:
```diff
@@ -27,8 +28,19 @@ function initializeWebViewClient(): void {
+			const intent = new android.content.Intent(android.content.Intent.ACTION_VIEW);
+			intent.setData(android.net.Uri.parse(url));
+
+			const packageManager = application.android.context.getPackageManager();
```
Then you can use it here:
```
const packageManager = ad.getApplicationContext().getPackageManager();
```

This will properly avoid the import of `application` in this file.

NathanWalker reviewed (changes requested):
You can also sign the CLA when get a chance, thank you.

prabudevarrajan commented:
@NathanWalker Updated as per your comments. Please merge accordingly

prabudevarrajan commented:
@cla-bot check

CatchABus commented:
I'm quite against this one. I believe we should enhance method `openUrl` to actually check for the existence of an activity that supports schema using package manager. Then, we could return the boolean result of `openUrl` itself in `shouldOverrideUrl`.

@prabudevarrajan I had plans to work on this one but I'm okay with your PR too if you pay more attention to requested changes.

prabudevarrajan commented:
> I'm quite against merging this one. We duplicate part of code that already resides in `openUrl`. I believe we should enhance method `openUrl` to actually check for the existence of an activity that supports schema using package manager. Then, we could return the boolean result of `openUrl` itself in `shouldOverrideUrl` (which should fix your reported problem).
> 
> @prabudevarrajan I had plans to work on this one but I'm okay with your PR too if you pay more attention to my requested changes. I could even create a PR tonight and see if you agree to my changes.
> 
> Note: Permission `QUERY_ALL_PACKAGES` is quite dangerous. Everything is fine as long as apps contain queries for the schemas they have support for.

@CatchABus  As per your comments, I have updated the [logic](https://github.com/NativeScript/NativeScript/pull/10148/commits/79ce1a521a23fc3f6829fa07337e0ef5a81dc399) accordingly. 

Regarding QUERY_ALL_PACKAGES, it's not recommended unless the user is aware of what he is doing. I have updated the [trace](https://github.com/prabudevarrajan/NativeScript/blob/master/packages/core/utils/index.android.ts#L31) log accordingly.

CatchABus commented on packages/core/utils/index.android.ts:
```diff
@@ -22,8 +22,16 @@ export function openUrl(location: string): boolean {
 		intent.addFlags(android.content.Intent.FLAG_ACTIVITY_NEW_TASK);
 
+		// Handle schemes like mailto, tel, etc
+		if (intent.resolveActivity(packageManager) == null) {
```
Since we bother calling `resolveActivity`, we can make use of its return value.
Something like:
```ts
const intentComponent = intent.resolveActivity(packageManager);

// Handle schemes like mailto, tel, etc
if (intentComponent == null) {
    return false;
}

intent.setComponent(intentComponent);
```
This is usually not needed because system will search and find the appropriate class if any, but now that we've already resolved it we can save a bit of work that would be done internally.
See: https://developer.android.com/reference/android/content/Intent#setComponent(android.content.ComponentName)

prabudevarrajan replied:
@CatchABus I believe we should allow the user to pick the appropriate apps for the intent. Setting setComponent() will restrict the app only to the intentComponent. resolveActivity() will always return only one activity compnent even if multiple activities match the intent. 

**resolveActivity()**
If multiple activities are found to satisfy the intent, the one with the highest priority will be used. If there are multiple activities with the same priority, the system will either pick the best activity based on user preference, or resolve to a system class that will allow the user to pick an activity and forward from there.
https://developer.android.com/reference/android/content/Intent#resolveActivity(android.content.pm.PackageManager)

**setComponent()**
If this class is defined, the specified class will always be used regardless of the other fields. You should only set this value when you know you absolutely want a specific class to be used; otherwise it is better to let the system find the appropriate class so that you will respect the installed applications and user preferences.

CatchABus replied:
The whole thing about `startActivity` is a bit of a mystery.
In `resolveActivity` details, you might also read this:
```
This API is called for you as part of starting an activity from an intent. You do not normally need to call it yourself.
```

Still, I'll agree with you that we shouldn't meddle with it more so let's leave it be.

farfromrefug commented:
I dont think QUERY_ALL_PACKAGES is necessary here or should even be mentioned. As @CatchABus said it is a dangerous perm in the sense that google is likely to refuse your app. The trace error should remove mention of it. It would wrongly induce users into thinking they need that permission when they dont

prabudevarrajan commented:
> I dont think QUERY_ALL_PACKAGES is necessary here or should even be mentioned. As @CatchABus said it is a dangerous perm in the sense that google is likely to refuse your app. The trace error should remove mention of it. It would wrongly induce users into thinking they need that permission when they dont

Yup agreed. I have updated the trace error [message](https://github.com/NativeScript/NativeScript/pull/10148/commits/9c6d3e0b9057f125d5a86fa88b49a20186c1d83f) accordingly.

CatchABus commented:
@prabudevarrajan Can you take a look at this one? https://github.com/NativeScript/NativeScript/pull/10148#discussion_r1059161592

I think we're good if it's done and works.

NathanWalker commented:
Nice job @prabudevarrajan !
